File: framework/languages/java/samples/java/ShoppingMall/Server/OrderWorkflow/src/main/java/systems/zlink/samples/shoppingmall/server/orderworkflow/spots/OrderWorkflowSpot.java
package systems.zlink.samples.shoppingmall.server.orderworkflow.spots;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import systems.zlink.framework.spots.ZLinkInstanceSpot;
import systems.zlink.framework.spots.ZLinkInstanceSpotContext;
import systems.zlink.samples.shoppingmall.server.orderworkflow.spots.handlers.ContinueOrderWorkflowSpotHandler;
import systems.zlink.samples.shoppingmall.server.orderworkflow.spots.handlers.PrepareInventoryReservedCheckpointSpotHandler;
import systems.zlink.samples.shoppingmall.server.orderworkflow.spots.handlers.RebuildOrderProjectionSpotHandler;
import systems.zlink.samples.shoppingmall.server.orderworkflow.spots.handlers.RunOrderWorkflowMsgHandler;
import systems.zlink.samples.shoppingmall.server.orderworkflow.spots.handlers.StartOrderWorkflowSpotHandler;
import systems.zlink.samples.shoppingmall.shared.contracts.Messages;

public final class OrderWorkflowSpot implements ZLinkInstanceSpot {
    private final ZLinkInstanceSpotContext context;

    public OrderWorkflowSpot(ZLinkInstanceSpotContext context) {
        this.context = context;
    }

    @Override
    public ZLinkInstanceSpotContext context() {
        return context;
    }

    // --8<-- [start:doc-sm-close-terminal]
    public CompletionStage<Void> closeIfTerminal(Messages.OrderState state) {
        if (Messages.OrderStatuses.Confirmed.equals(state.status())
            || Messages.OrderStatuses.Failed.equals(state.status())) {
            return context.close().thenApply(closed -> null);
        }
        return CompletableFuture.completedFuture(null);
    }
    // --8<-- [end:doc-sm-close-terminal]

}

File: framework/languages/java/samples/kotlin/ShoppingMall/Server/OrderWorkflow/src/main/kotlin/systems/zlink/samples/kotlin/shoppingmall/server/orderworkflow/OrderWorkflowSpot.kt
package systems.zlink.samples.kotlin.shoppingmall.server.orderworkflow

import systems.zlink.framework.spots.ZLinkInstanceSpot
import systems.zlink.framework.spots.ZLinkInstanceSpotContext
import systems.zlink.samples.kotlin.shoppingmall.server.orderworkflow.handlers.ContinueOrderWorkflowHandler
import systems.zlink.samples.kotlin.shoppingmall.server.orderworkflow.handlers.PrepareInventoryReservedHandler
import systems.zlink.samples.kotlin.shoppingmall.server.orderworkflow.handlers.RebuildOrderProjectionHandler
import systems.zlink.samples.kotlin.shoppingmall.server.orderworkflow.handlers.StartOrderWorkflowHandler

class OrderWorkflowSpot(
    private val instanceContext: ZLinkInstanceSpotContext,
) : ZLinkInstanceSpot {
    override fun context(): ZLinkInstanceSpotContext = instanceContext

    fun requireOrder(orderId: String) {
        require(orderId == instanceContext.spotId()) {
            "request order does not match workflow Spot: $orderId"
        }
    }
}
